### PR TITLE
Bodypart received damage is run through armor check in certain situations

### PR DESCRIPTION
**Pull request by yoyobatty**
[Changelogs]: # (Your PR should contain a detailed changelog of notable changes, titled and categorized appropriately. This includes, new features, sprites, sounds, balance changes, admin tools, map edits, removals, big refactors, config changes, hosting changes and important fixes. An example changelog has been provided below for you to edit. If you need additional help, read https://github.com/tgstation/tgstation/wiki/Changelogs)

:cl: YoYoBatty
tweak: Body part damage is now run through armor check in certain situations.
/:cl:

[why]: # (Please add a short description [two lines down] of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.)

Stops you from taking 20 fucking brute damage anytime you collide with a wall even with a DS suit on.

### CODE REVIEW

**Comment by ATH1909** [comment]
What in all will this affect? Will this affect chems that deal damage? Fire? Acid? Hitting yourself into a wall with a baseball bat?

**Comment by Shadowflame909** [comment]
Buff?

**Comment by necromanceranne** [comment]
Massive buff in small ways.

**Comment by ATH1909** [comment]
This will make getting 100%+ resistance to melee damage much more desirable/useful/interesting, which I approve of.

**Comment by yoyobatty** [comment]
It's just to stop damage when you get thrown into wall hard bypassing armor and therefore doing 20 dmg

File: code/modules/mob/living/carbon/damage_procs.dm [hunk 1]
```diff
@@ -153,7 +153,7 @@
 	if(!parts.len)
 		return
 	var/obj/item/bodypart/picked = pick(parts)
-	if(picked.receive_damage(brute, burn, stamina))
+	if(picked.receive_damage(brute, burn, stamina,run_armor_check(picked, brute?"melee":burn?"fire":stamina?"bullet":null)))
```
Comment (Cyberboss) [inline code/modules/mob/living/carbon/damage_procs.dm]: Parenthesis and spaces please

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by yoyobatty** [comment]
Done @cyberboss

**Comment by ExcessiveUseOfCobblestone** [comment]
good pr

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 2]
```diff
@@ -142,14 +142,17 @@
 //Applies brute and burn damage to the organ. Returns 1 if the damage-icon states changed at all.
 //Damage will not exceed max_damage using this proc
 //Cannot apply negative damage
-/obj/item/bodypart/proc/receive_damage(brute = 0, burn = 0, stamina = 0, updating_health = TRUE, required_status = null)
+/obj/item/bodypart/proc/receive_damage(brute = 0, burn = 0, stamina = 0, blocked = FALSE, updating_health = TRUE, required_status = null)
+	var/hit_percent = (100-blocked)/100
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/surgery/bodyparts/bodyparts.dm]: here blocked is an integer, but on L145 it's a boo

File: code/modules/mob/living/carbon/damage_procs.dm [hunk 3]
```diff
@@ -153,7 +153,7 @@
 	if(!parts.len)
 		return
 	var/obj/item/bodypart/picked = pick(parts)
-	if(picked.receive_damage(brute, burn, stamina))
+	if(picked.receive_damage(brute, burn, stamina,run_armor_check(picked, (brute ? "melee" : burn ? "fire" : stamina ? "bullet" : null))))
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/damage_procs.dm]: > armor doesn't use defines

rip

**Comment by Shadowflame909** [comment]
Colossus bullets intentionally ignores armor damage. Will this be an indirect buff to that?

File: code/modules/mob/living/carbon/damage_procs.dm (hunk 3 above)
Comment (yoyobatty) [reply code/modules/mob/living/carbon/damage_procs.dm]: Unfortunately not

File: code/modules/surgery/bodyparts/bodyparts.dm (hunk 2 above)
Comment (yoyobatty) [reply code/modules/surgery/bodyparts/bodyparts.dm]: Just copied how other code did it, fixed.

**Comment by yoyobatty** [comment]
If you want to see what this affects, here you go: https://github.com/tgstation/tgstation/search?q=take_bodypart_damage&unscoped_q=take_bodypart_damage

**Review by Cyberboss — CHANGES_REQUESTED** [review]
There are several things in that search that should ignore armor like diseases, magic, nanites, etc...

**Comment by ExcessiveUseOfCobblestone** [comment]
Some of those would be a correct assumption

**Comment by yoyobatty** [comment]
Would you prefer I have those send a different dam type or should I remove the default assumed dam type in this proc or both ?

**Comment by ExcessiveUseOfCobblestone** [comment]
I'd look at where this proc is called and just set those to not check for armor.

Things that bypass armor should be an exception anyways.

**Comment by yoyobatty** [comment]
Okay, it now has a check defaulted to off, it's set to check when you accidentally hit yourself with an esword (clumsiness trait required) and when colliding with a wall or someone else.

**Comment by Cyprex** [comment]
We could have had a major stealth buff if only you didn't ask shadowflame...

File: code/modules/mob/living/carbon/damage_procs.dm [hunk 4]
```diff
@@ -148,12 +148,12 @@
...
 //It automatically updates damage overlays if necessary
 //It automatically updates health status
-/mob/living/carbon/take_bodypart_damage(brute = 0, burn = 0, stamina = 0, updating_health = TRUE, required_status)
+/mob/living/carbon/take_bodypart_damage(brute = 0, burn = 0, stamina = 0, updating_health = TRUE, required_status, check_armor = FALSE)
 	var/list/obj/item/bodypart/parts = get_damageable_bodyparts(required_status)
 	if(!parts.len)
 		return
 	var/obj/item/bodypart/picked = pick(parts)
-	if(picked.receive_damage(brute, burn, stamina))
+	if(picked.receive_damage(brute, burn, stamina,check_armor ? run_armor_check(picked, (brute ? "melee" : burn ? "fire" : stamina ? "bullet" : null)) : 0))
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/damage_procs.dm]:
```suggestion
	if(picked.receive_damage(brute, burn, stamina,check_armor ? run_armor_check(picked, (brute ? "melee" : burn ? "fire" : stamina ? "bullet" : null)) : FALSE))
```

**Pull request merged**